presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkMapFlatWriter.java

```diff
@@ -123,64 +129,125 @@ private void doWrite(BenchmarkData data)
+        // It has 11 elements. The element position is the frequency bucket.
+        // The value of the element is the weight indicating how many keys should
+        // fall into this frequency bucket. All weights will be normalized before usage.
+        // The last element is the weight for keys with 100% frequency.
+        //
+        // Illustration:
+        // [this many keys (after weights normalization) come with frequency <10%, ... with frequency <20%, ..., ...with frequency 100%]
+        //
+        // For example:
+        // [1, 1, 0,0,0,0,0,0,0,0, 1] means that 1/3 of keys would come with 10% frequency, another 1/3 with 20% frequency, and the last 1/3 with 100%.
+        @Param("12, 2, 5, 3, 4, 5, 2, 10, 7, 45, 0")
+        private String frequencyHistogram = "12, 2, 5, 3, 4, 5, 2, 10, 7, 45, 0";
+
         private List<Block> blocks;
 
         @Setup
         public void setup()
                 throws Exception
         {
-            FunctionAndTypeManager testFunctionAndTypeManager = createTestFunctionAndTypeManager();
-            Type keyType = testFunctionAndTypeManager.getType(parseTypeSignature(this.keyType));
-            Type valueType = testFunctionAndTypeManager.getType(parseTypeSignature(this.valueType));
+            Random random = new Random(SEED);
+
+            FunctionAndTypeManager typeManager = createTestFunctionAndTypeManager();
+            this.type = (MapType) typeManager.getType(parseTypeSignature(typeSignature));
+            Type keyType = type.getKeyType();
+            Type valueType = type.getValueType();
 
-            this.type = mapType(keyType, valueType);
             this.temporaryDirectory = createTempDir();
             this.orcFile = new File(temporaryDirectory, randomUUID().toString());
+            this.orcFile.deleteOnExit();
 
             ImmutableList.Builder<Block> blocks = ImmutableList.builder();
 
-            List<Long> keys = new ArrayList<>();
+            List<Long> keys = new ArrayList<>(distinctKeyCount);
             for (int i = 0; i < distinctKeyCount; i++) {
                 keys.add((long) i);
             }
+            Collections.shuffle(keys, random);
+
+            List<Integer> weights = Arrays.stream(frequencyHistogram.split(","))
+                    .map(String::trim)
+                    .map(Integer::valueOf)
+                    .collect(Collectors.toList());
+            assertEquals(weights.size(), 11, "Number of weights is expected to be 11");
+
+            double weightSum = weights.stream().mapToInt(Integer::intValue).sum();
+            double[] keyFrequencies = new double[distinctKeyCount];
+            int offset = 0;
+            for (int i = 0; i < weights.size(); i++) {
+                double frequency = 0.1 * (i + 1);
+                int keyCount = (int) (distinctKeyCount * weights.get(i) / weightSum);
+                for (int j = 0; j < keyCount; j++) {
+                    keyFrequencies[j + offset] = frequency;
```

Comment:
nit: set keyFrequencies[offset++] = frequency;

After the outer for loop, add checkState(offset == distinctKeyCount), after reading the code few times, I understand the logic now :)

Reply:
Done. Instead of checkState, I'm adjusting down the keyCount.